File: src/Tests/Performance/PerfString/PerfString.cs
namespace go;

using fmt = fmt_package;
using time = time_package;

partial class main_package {

internal static nint run(nint n) {
    nint total = 0;
    var buf = new slice<byte>(0, 32);
    for (nint i = 0; i < n; i++) {
        buf = buf[..0];
        nint v = i;
        if (v == 0) {
            buf = append(buf, (byte)((rune)'0'));
        }
        while (v > 0) {
            buf = append(buf, (byte)((rune)'0' + v % 10));
            v /= 10;
        }
        @string s = ((@string)buf);
        total += len(s);
        total += (nint)s[0];
        if (i % 1000 == 0) {
            @string t = s + "-"u8 + s;
            total += len(t);
            total += (nint)t[len(t) - 1];
        }
    }
    return total;
}

internal static void Main() {
    var start = time.Now().UnixNano();
    nint total = run(10000000);
    var elapsed = time.Now().UnixNano() - start;
    fmt.Println("checksum:", total);
    fmt.Println("elapsed_ns:", elapsed);
}

} // end main_package

File: src/Tests/Performance/PerfString/PerfString.go
package main

import (
	"fmt"
	"time"
)

// String building and conversion: measures byte-slice append, []byte->string
// conversion, string indexing, and concatenation (@string emulation in C#).

func run(n int) int {
	total := 0
	buf := make([]byte, 0, 32)

	for i := 0; i < n; i++ {
		buf = buf[:0]
		v := i
		if v == 0 {
			buf = append(buf, '0')
		}
		for v > 0 {
			buf = append(buf, byte('0'+v%10))
			v /= 10
		}

		s := string(buf)
		total += len(s)
		total += int(s[0])

		if i%1000 == 0 {
			t := s + "-" + s
			total += len(t)
			total += int(t[len(t)-1])
		}
	}

	return total
}

func main() {
	start := time.Now().UnixNano()

	total := run(10000000)

	elapsed := time.Now().UnixNano() - start
	fmt.Println("checksum:", total)
	fmt.Println("elapsed_ns:", elapsed)
}
